### ISSUE TITLE
Large spike in database read operations during bulk event deletion

### ISSUE THREAD

**Issue by tcdowney** [cloudfoundry/cloud_controller_ng#1373]
_Originally posted by @abg in https://github.com/cloudfoundry/cloud_controller_ng/issues/1339#issuecomment-501323530 as a followup to our fix. Since this change has been out in the past several capi-releases, I'm moving this out to a separate issue to make it easier for us to track._

**Original issue:** https://github.com/cloudfoundry/cloud_controller_ng/issues/1339

---

We've seen mixed result from this change.  The "stop the world" effect we previously observed under Galera clusters has disappeared which has been a great improvement.   However, this has been replaced by a large spike in read operations in the storage engine over a few hour window:

 ![innodb_rw_s](https://user-images.githubusercontent.com/163206/59284068-3e6b0d00-8c31-11e9-8c3f-0cfad4c53948.png)

Digging into this a bit, we see the chunking query is not executed efficiently by MySQL.   The current implementation constructs a query of the form:

```
DELETE FROM `events` WHERE (`id` IN (SELECT * FROM (SELECT `id` FROM `events` WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000) AS `t1`));
```

From the mysql slow query log it is obvious this query is doing a full table scan:

<pre>
# Query_time: 60.012630 Lock_time: 0.000076 Rows_sent: 0 <b>Rows_examined: 20913373</b> Rows_affected: 1000
DELETE FROM events WHERE (id IN (SELECT * FROM (SELECT id FROM events WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000) AS t1));
</pre>

For this environment, there are a few hundred thousand events to cleanup daily, so this gets run a few hundred times over the course of ~3 hours.   The event cleanup process is essentially monopolizing a full CPU core during this time and thrashes the InnoDB buffer pool needlessly.

**Recommendations**

The load from the current implementation can cause problems in some environments.

There are a few ways to rewrite this query to obtain a more efficient access pattern.  The most obvious would be the simple `DELETE ... LIMIT` form that MySQL supports:

```
DELETE FROM events WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000
```

MySQL is able to use an index scan in this case and efficiently finds the chunk of rows to delete.  This form is MySQL specific and not generalizable to other databases.   Postgres works fine with the current form.

Alternatively a small tweak to the current query can result in a much improved query plan:

<pre>
DELETE events <b>FROM events</b> WHERE (id IN (SELECT * FROM (SELECT id FROM events WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000) AS t1))
</pre>

This form uses a MySQL [multiple-table delete trick](https://mysqlserverteam.com/multi-table-trick/) to kick the MySQL optimizer into generating a better access plan.

Either of these forms execute in dozens of milliseconds rather than ~60s per chunk.

**Comment by tcdowney** [comment]
@abg do you know why this is doing a full table scan?

```
# Query_time: 60.012630 Lock_time: 0.000076 Rows_sent: 0 Rows_examined: 20913373 Rows_affected: 1000
DELETE FROM events WHERE (id IN (SELECT * FROM (SELECT id FROM events WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000) AS t1));
```

If we executed the two parts to that query separately do you think it would improve things? Some rough psuedo-SQL for what I mean below:

Grab the ids we want from:
`SELECT * FROM (SELECT id FROM events WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000`

Toss those ids into:
`DELETE FROM events WHERE id in ids;`

We're currently building this query up in our ORM here:
https://github.com/cloudfoundry/cloud_controller_ng/blob/c31fd2e103dff84288c8aab0dff7bf3096fdb3a1/lib/database/batch_delete.rb#L22

What that's doing is nesting a Sequel (our ORM) [dataset](https://sequel.jeremyevans.net/rdoc/classes/Sequel/Dataset.html) within a larger one and generating this nested query.

I'm wondering if this would be better (this time some Ruby pseudocode 😛) :
```
ids = set.select(:id) # create an array of ids in Ruby
dataset.model.where(id: ids).delete # filter on the list of ids instead
```

**Comment by tcdowney** [comment]
> I do see the attractiveness in the simplicity of implementation by just generating a big IN list.

Yeah, since we support running against both Postgres and MySQL I'd prefer something that would work reasonably well in both. We actually did try the `DELETE FROM events WHERE (created_at < '2019-03-10 20:00:00') LIMIT 1000` style delete originally until we realized it wasn't portable.

**Comment by tcdowney** [comment]
@abg we implemented these changes in [#166644422](https://www.pivotaltracker.com/story/show/166644422) and it shipped in capi-release `1.83.0`. Have you noticed any improvements in your environment?

**Comment by cwlbraa** [comment]
Closing due to unanswered question. I think we fixed this.
